File: src/main/java/io/xol/chunkstories/core/item/armor/ItemArmor.java
//
// This file is a part of the Chunk Stories Core codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.core.item.armor;

import io.xol.chunkstories.api.item.Item;
import io.xol.chunkstories.api.item.ItemDeclaration;

import java.util.Collection;

public abstract class ItemArmor extends Item {
    private double damageMultiplier = 0.5;
    private String overlay = "notexture";

    public ItemArmor(ItemDeclaration type) {
        super(type);
    }

    /**
     * Returns either null (and affects the entire holder) or a list of body parts
     * it cares about
     */
    public abstract Collection<String> bodyPartsAffected();

    /**
     * Returns the multiplier, optional bodyPartName (might be null, handled that
     * case) for entities that support it
     */
    public float damageMultiplier(String bodyPartName) {
        return (float) damageMultiplier;
    }

    public String getOverlayTextureName() {
        return overlay;
    }

}

File: src/main/java/xyz/chunkstories/core/item/armor/ItemArmor.kt
//
// This file is a part of the Chunk Stories Core codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.core.item.armor

import xyz.chunkstories.api.content.json.asFloat
import xyz.chunkstories.api.content.json.asString
import xyz.chunkstories.api.item.Item
import xyz.chunkstories.api.item.ItemDefinition

abstract class ItemArmor(definition: ItemDefinition) : Item(definition) {
    private val mutiplier: Float
    val overlayTextureName: String

    init {
        mutiplier = definition["damageMultiplier"].asFloat ?: 0.5f
        overlayTextureName =definition["overlay"].asString ?: "notexture"
    }

    /**
     * Returns either null (and affects the entire holder) or a list of body parts
     * it cares about
     */
    abstract fun bodyPartsAffected(): Collection<String>

    /**
     * Returns the multiplier, optional bodyPartName (might be null, handled that
     * case) for entities that support it
     */
    fun damageMultiplier(bodyPartName: String?): Float {
        return mutiplier
    }

}
